src/functions.cpp

```diff
@@ -0,0 +1,204 @@
+  }
+  str += ")";
+  return str;
+}
+} // namespace
+
+FunctionRegistry::FunctionRegistry(std::ostream &out) : out_(out)
+{
+}
+
+Function &FunctionRegistry::add(Function::Origin origin,
+                                std::string_view name,
+                                const SizedType &returnType,
+                                std::vector<Param> params)
+{
+  return add(origin, {}, name, returnType, params);
+}
+
+Function &FunctionRegistry::add(Function::Origin origin,
+                                std::string_view ns,
+                                std::string_view name,
+                                const SizedType &returnType,
+                                std::vector<Param> params)
+{
+  all_funcs_.push_back(std::make_unique<Function>(
+      origin, std::string{ name }, returnType, params));
+  Function &newFunc = *all_funcs_.back().get();
+  std::string fq_name;
+  if (ns.empty())
+    fq_name = name;
+  else
+    fq_name = std::string{ ns } + "::" + newFunc.name();
+
+  // Check for duplicate function definitions
+  // The assumption is that builtin functions are all added to the registry
+  // before any user-defined functions.
+  for (const Function &func : funcs_by_fq_name_[fq_name]) {
+    if (func.origin() == Function::Origin::Builtin)
+      continue; // Allow duplicates of builtins
+    if (func.params().size() != params.size())
+      continue;
+    bool is_dup = true;
+    for (size_t i = 0; i < func.params().size(); i++) {
+      if (func.params()[i].type() != params[i].type()) {
+        is_dup = false;
+        break;
+      }
+    }
+    if (is_dup) {
+      throw std::runtime_error("Cannot redefine existing function: " + fq_name +
+                               param_types_str(params));
+    }
+  }
+
+  funcs_by_fq_name_[fq_name].push_back(newFunc);
+  return newFunc;
+}
+
+namespace {
+bool can_implicit_cast(const SizedType &from, const SizedType &to)
+{
+  if (from.FitsInto(to))
+    return true;
+
+  if (from.IsStringTy() && to.IsPtrTy() && to.GetPointeeTy()->IsIntTy() &&
+      to.GetPointeeTy()->GetSize() == 1) {
+    // Allow casting from string to int8* or uint8*
+    return true;
+  }
+
+  // Builtin and script functions do not care about string sizes. External
+  // functions cannot be defined to accept string types (they'd take char*)
+  if (from.IsStringTy() && to.IsStringTy())
+    return true;
+
+  return false;
+}
+} // namespace
+
+/*
+ * Find the best function called 'name' for the given argument types.
+ *
+ * Returns either a single function or nullptr, when no such function exists or
+ * a single best match can not be found.
+ *
+ * Summary of overload resolution rules:
+ * 1. If a single valid candidate function exists, pick that
+ * 2. If one of the candidates is an exact type match, pick that
+ * 3. If multiple candidates are exact type matches, prefer script/external
+ *    functions over builtin functions
+ * 4. If there are multiple, non-exact matches, return no-match
```

Comment:
Can we punt on function overloading? I am very much not a fan of function overloading. That aside, I think it'd be better to add it later and after some more discussion. Nothing on the project roadmap needs it yet.